src/core/DY.Core.Label.fst

```diff
@@ -380,3 +292,166 @@ let join_flow_to_public_eq tr x1 x2 =
   reveal_opaque (`%can_flow) (can_flow);
   reveal_opaque (`%join) (join);
   reveal_opaque (`%public) (public)
+
+/// TODO comment
+
+[@@"opaque_to_smt"]
+val state_pred_label:
+  (principal -> state_id -> bytes -> prop) ->
+  label
+let state_pred_label p = mk_label {
+  is_corrupt = (fun tr ->
+    exists time prin sess_id content.
+      event_exists tr (Corrupt time) /\
+      event_at tr time (SetState prin sess_id content) /\
+      p prin sess_id content
+  );
+  is_corrupt_later = (fun tr1 tr2 -> ());
+}
+
+val state_pred_label_pred_implies:
+  (principal -> state_id -> bytes -> prop) ->
+  (principal -> state_id -> bytes -> prop) ->
+  prop
+let state_pred_label_pred_implies p1 p2 =
+  forall p s c. p2 p s c ==> p1 p s c
```

Comment:
The order of `p1` and `p2` is confusing.
I would have expected it to read like ```p1 `implies` p2```

Reply:
Indeed, this is to match the direction of `can_flow` (which is reversed) https://github.com/REPROSEC/dolev-yao-star-extrinsic/blob/9c8d4d8d300755ea6d338c8afd147ff27ba1fa24/src/core/DY.Core.Label.fst#L116-L118
This ensures the lemma about `can_flow` and `state_pred_label` isn't weird https://github.com/REPROSEC/dolev-yao-star-extrinsic/blob/741561406bc3aa2644274c4193c60ccb5c234e63/src/core/DY.Core.Label.fst#L323-L324

Maybe a better name would be `state_pred_label_pred_can_flow`?
Overall I agree the names are not great -- I procrastinated finding good names to focus on trying to make the technique work, now that all the predicates and functions are there we should certainly do a holistic renaming before merging.

Reply:
I changed the name of the lemma to `state_pred_label_can_flow` and did a pass on the naming of the various labels